### DISCUSSION TITLE
https://github.com/tokio-rs/tokio/issues/2316

### DISCUSSION THREAD

**Discussion by avnerbarr** [tokio-rs/tokio#4424] [Q&A]
similar to https://github.com/tokio-rs/tokio/issues/2316

test code is as follows:

This runs as expected
```
#[test]
fn test_timers() {
    let mut rt = tokio::runtime::Builder::new_multi_thread()
        .enable_all()
        .enable_time()
        .build()
        .unwrap();

    rt.block_on(async {
        println!("waiting");
        let _ = tokio::time::sleep(std::time::Duration::from_secs(3));
        println!("finished waiting");
    });
}
```

But with timers if they expire it will crash for me:
```
#[test]
fn test_timers() {
    let mut rt = tokio::runtime::Builder::new_multi_thread()
        .enable_all()
        .enable_time()
        .build()
        .unwrap();

// this crashes
    let f = tokio::time::timeout(std::time::Duration::from_secs(1), async {
        println!("waiting");
        let _ = tokio::time::sleep(std::time::Duration::from_secs(3));
        println!("finished waiting");
    });
    rt.block_on(f);
}

there is no reactor running, must be called from the context of a Tokio 1.x runtime
thread 'a::test_b' panicked at 'there is no reactor running, must be called from the context of a Tokio 1.x runtime'
```

Example:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7b24d2085a3127c2e7f073091c90eff7

**Comment by avnerbarr** [comment]
maybe this explains? what is the workaround?

```
/// It can also panic whenever a timer is created outside of a
/// Tokio runtime. That is why `rt.block_on(sleep(...))` will panic,
/// since the function is executed outside of the runtime.
/// Whereas `rt.block_on(async {sleep(...).await})` doesn't panic.
/// And this is because wrapping the function on an async makes it lazy,
/// and so gets executed inside the runtime successfully without
/// panicking.
#[track_caller]
pub fn timeout<T>(duration: Duration, future: T) -> Timeout<T>
where
    T: Future,
{
```

**Comment by Darksonn** [comment]
This is because your timer is created outside of the runtime. This works:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=21d7db1919bafbd3ea75cacefd3fef0a

**Comment by avnerbarr** [comment]
Yes! of course after posting this question figured it out (had to go through the stack trace and read the documentation of the internals!)

Maybe we can add a check that this is the case and have an easy to read message? 

Another issue I have now (unrelated possibly) is that the test will complete but then I will get a message like: 

```
thread 'tokio-runtime-worker' panicked at 'Cannot drop a runtime in a context where blocking is not allowed. This happens when a runtime is dropped from within an asynchronous context.', /tokio-1.15.0/src/runtime/blocking/shutdown.rs:51:21
```

**Comment by avnerbarr** [comment]
(I have a spawned task in the function being tested + a timed task)
pseudo:
```
fn work(&self) -> Result {
    self.rt.spawn( async work in background);
    // but return some intermediate result with a timeout per the question you have answered
    self.rt.block_on(get_with_timeout per the question above...)
}

#[test]
fn test_work() {
  let x = ...;
   x.work();
}
```

**Comment by Darksonn** [comment]
> Cannot drop a runtime in a context where blocking is not allowed.

You are dropping a runtime from inside a runtime. Did you create two runtimes?

**Comment by avnerbarr** [comment]
It might have been created deeper inside. Is there some constraint on how many runtimes to have? Should runtimes be shared across modules / crates? if so how ?

**Comment by Darksonn** [comment]
There's no constraint on the number of runtimes, but you can only use one runtime at the time on a single OS thread. It is generally more performant to share a single runtime, and most projects should not create more than one.

**Reply by avnerbarr** [reply]
I have a large legacy code base that is sync (standard rust) and some newer code is running async. how would you suggest bridging this? Is there a way for an `async fn` to access the current runtime (i.e. `Tokio::thread::get_rt?`)? 

Notice that the newer components have "sync" modes (sync version of the functions) which create an internal runtime and blocks by calling the async versions. 

I assume that in some situations (`fn` closures or similar effects) I could have nested runtimes (this is hard to test on a large code base with complicated dependencies) .

what is the suggested way to share a runtime in this case (gradual migration)? 

for a concrete toy example:

```
fn main() {
   // legacy
   work()
}
fn work() {
      // patched to work with new functionality
      let another_rt = tokio::runtime::Builder::new_multi_thread()
                .enable_all()
                .build()
                .unwrap();

    // want to run both in parallel
    // sync code which calls out to async code with an internal rt
     new_thing.helper(timeout_duration); // this has an internal runtime (I suppose it could be refactored to async?
    // this is standard blocking so should be wrapped in async?
     old_thing.legacy_work();
    
/// modify it to perform in the runtime (pseudo)
   // now we have a runtime in a runtime (unless share runtime somehow or modify function to async?)
    let handle1 = another_rt.spawn(async { new_thing.helper(timeout)});
    let handle2 = another_rt.spawn(async {
       old_thing.legacy_work();
    };




}

fn legacy_work() {
       /// does some work but also possibly calls something with a different runtime
}

fn helper1(timeout: Duration) {
  let (s, r) = tokio::sync::oneshot::channel();
        // from another mod
        let cloned = self.client.clone();
        let async_work = async move {
            let x = cloned.work().await;
            let _ = s.send(x);
        };

        self.rt.spawn(async_work); // tokio::runtime::Builder::new_multi_thread()
                .enable_all()
                .build()
                .unwrap(),

        // allow timeout
        let future = async { tokio::time::timeout(*time_limit, r).await };
        self.rt
            .block_on(future)
            .ok()
            .map(|x| x.ok().flatten())
            .flatten()
            .unwrap_or_default()
    }

}


// a different crate
struct Client {
  runtime: Runtime
}

impl Client {
   async fn work(&self) {...}
   fn work_sync(&self) { self.runtime.block_on(async work);}
}
```

**Reply by Darksonn** [reply]
> Is there a way for an async fn to access the current runtime?

Yes. This is [`tokio::runtime::Handle::current()`](https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.current).

---

In general, if you are in async code, then you are not supposed to call a non-async function if that function call could block. This includes any function that creates a runtime and uses `block_on`, but it also includes calls that block for other reasons (e.g. calls `std::thread::sleep`). The correct way to call such a function from async code is to wrap the call in [`spawn_blocking`](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html). If you make sure to do this, then nesting runtimes is ok, because using `block_on` inside `spawn_blocking` is perfectly fine.

Check out [this article][1] for more info.

[1]: https://ryhl.io/blog/async-what-is-blocking/

**Reply by avnerbarr** [reply]
> This includes any function that creates a runtime and uses block_on
If I understand correctly that raises the question of how would some code know if another piece of code which relies on it (calling it) is async? (i.e. would cause this to be blocking), does that mean I should always test for `tokio::runtime::Handle::current()` before spawning to a runtime from within a sync fn?

to illustrate what I mean:
```
fn sync_blocking(&self) {
    // should I be doing this???
   if tokio::runtime::Handle::current() {
        tokio::runtime::Handle::current().block_on(work());
   } else {
      self.runtime.block_on(work());
   }
   
}

async fn work() {
}

```

**Reply by Darksonn** [reply]
It is usually incorrect to _check_ whether there is a current runtime. Either you should require that your code is called in the context of a runtime and do `Handle::current().block_on(...)`, or you should always use your own runtime. I note that `Handle::current().block_on(...)` will only work inside `spawn_blocking` calls, whereas creating your own runtime would also work on custom threads spawned with `std::thread::spawn`.

**Reply by avnerbarr** [reply]
when you say

> require that your code is called in the context of a runtime and do Handle::current().block_on(...)

How do I require that? does that mean "when inside an "async fn" function, and must block (wouldn't I just do `await`?), use `Handle::current().block_on(...)`" . not sure how a function would know it is in a context of a runtime otherwise?.

or you should always use your own runtime
How do I make a 3rd party use my runtime?


consider this example (real world):


```
// blocking function calling async function calling blocking function calling async ... :)
fn do_work() {
    let rt = tokio::runtime::Builder::new_multi_thread()
        .enable_all()
        .build()
        .unwrap();
// foo is async for various reasons, in this toy scope it is not obvious
    rt.block_on(work2()); // 
}


async fn work2() {
   let b = LegacyBridge::new();
    // f1 is blocking , there are other things in this function which are called .await but removed for clarity
   b.work_with_timeout(); 
}


struct LegacyBridge {
  rt: tokio::runtime::Runtime,
  client: SomeFancyAsyncNewCodeClient // has async fn inside it (and belongs to another crate)
}

impl LegacyBridge {

  // performs what is perceived to be blocking work with a timeout and returns vector of results
  // if it timed out the vector that is returned is empty, but the underlying work task should still be running, this is critical because it caches data for subsequent calls)
  fn work_with_timeout() -> Vec<Result> {
        // how would you suggest timing out this async work? I figured this solution, possibly there is more idiomatic mechanism?
        let (s, r) = tokio::sync::oneshot::channel();
        let cloned = self.client.clone();
        // want this function to timeout if it must but also should appear as blocking to legacy callers
        // block on the channel or timeout
        let task_async = async move {
           // the actual work which must be performed (network calls etc.)
            let x = clone.work(&*args).await;
            let _ = s.send(x);//  move the result to the blocking portion
        };


        // may take longer then timeout limit but want it to complete in background

        let _ = self.rt.spawn(task_async);



        // we need an extra wrap since tokio timeout must be created itself inside an executor and this function is a standard sync (not async) function so must do that explicitly by wrapping in an async block
        let future = async { tokio::time::timeout(*time_limit, r).await };
        // self.rt
.block_on(future)
            .ok()
            .map(|x| x.ok().flatten()) // this is just transformation to required result , not interesting for this example
            .flatten()
            .unwrap_or_default() // empty vector if error/timeout etc., not interesting why this may have occured for this example
  }
}
```

**Reply by Darksonn** [reply]
You should be doing this:
```Rust
async fn work2() {
    // Wrap any use of the blocking Legacy object in `spawn_blocking`.
    tokio::task::spawn_blocking(|| {
        let b = LegacyBridge::new();
        b.work_with_timeout();
    }).await.unwrap();
}
```

**Reply by avnerbarr** [reply]
👍  anyway to have a compile time check for this? I only now understand that I can uncover these issues at runtime, and that can lead to regressions no? (untested changes could crash at runtime with errors like: ```"Cannot start a runtime from within a runtime. This happens \
            because a function (like `block_on`) attempted to block the \
            current thread while the thread is being used to drive \
            asynchronous tasks."```

```
)

**Reply by Darksonn** [reply]
Unfortunately there's no compile-time check for this. There are several ideas for how to add something like that to the language, but none of them are implemented.

**Reply by avnerbarr** [reply]
How would you go about running async code in parallel with sync code from legacy sync code?
for example:

```
// sync function
fn do_work_in_parallel(&self) -> Vec<String> {
   // want to run do_work and do_work_async in parallel, combining the results 
// should I create a runtime each time the method is called (or perhaps an instance runtime? I have many of these objects possibly) vs should I use tokio::task::spawn_blocking (or some variation?) ... etc. etc.
   // what I want is w1 and w2 in parallel and then to wait and combine the results
   let w1 = do_work();  
   let w2 = do_work_async(); 
  return  w1 + w2
}

// the legacy sync work
fn do_work(&self) -> Vec<String>{
}

async fn do_work_async(&self) -> Vec<String>{
}
```

**Reply by Darksonn** [reply]
You can do that like this:
```Rust
fn do_work_in_parallel(&self) -> Vec<String> {
    let w1_handle = spawn_blocking(|| do_work());
    let w2 = do_work_async().await;
    let w1 = w1_handle.await.unwrap();
    return w1 + w2;
}
```
